ion-hash/src/type_qualifier.rs

```diff
@@ -30,55 +35,27 @@ impl TypeQualifier {
     /// laid out in the spec. In many cases, the `T` is determined by the Ion
     /// binary type code.
     ///
-    /// As much as possible, we use `const fn` to implement lookups logically.
-    /// The code is authored to read like the spec, without sacrificing
-    /// performance.
+    /// This method simply calls the appropriate `type_qualifier_$ion_type`
+    /// method.
     pub(crate) fn from_element<E>(elem: &E) -> TypeQualifier
     where
         E: Element + ?Sized,
     {
-        use IonTypeCode::*;
         match elem.ion_type() {
-            IonType::Null => combine(NullOrWhitespace, 0x0F),
-            IonType::Boolean => {
-                let q = match elem.as_bool() {
-                    None => 0x0F,
-                    Some(b) => match b {
-                        false => 0x00,
-                        true => 0x01,
-                    },
-                };
-                combine(Boolean, q)
-            }
-            IonType::Integer => {
-                let any = elem.as_any_int();
-                let t = match is_anyint_positive(any) {
-                    true => IonTypeCode::PositiveInteger,
-                    false => IonTypeCode::NegativeInteger,
-                };
-                let q = match any {
-                    Some(_) => 0x00,
-                    None => 0x0F,
-                };
-                combine(t, q)
-            }
-            IonType::Float => combine(Float, 0x00),
+            IonType::Null => type_qualifier_null(),
+            IonType::Boolean => type_qualifier_boolean(elem.as_bool()),
+            IonType::Integer => type_qualifier_integer(elem.as_any_int()),
+            IonType::Float => type_qualifier_float(elem.as_f64()),
             /*IonType::Decimal => Decimal,
-            IonType::Timestamp => Timestamp,
-            IonType::Symbol => Symbol,*/
-            IonType::String => {
-                let q = match elem.as_str() {
-                    Some(_) => 0x00,
-                    None => 0x0F,
-                };
-                combine(String, q)
-            }
-            /*IonType::Clob => Clob,
-            IonType::Blob => Blob,
-            IonType::List => List,
-            IonType::SExpression => SExpression,*/
-            /*IonType::Struct => combine(Struct, 0),*/
-            _ => todo!(),
+            IonType::Timestamp => Timestamp,*/
+            IonType::Symbol => type_qualifier_symbol(elem.as_sym()),
+            IonType::String => type_qualifier_string(elem.as_str()),
+            IonType::Clob => type_qualifier_clob(elem.as_bytes()),
+            IonType::Blob => type_qualifier_blob(elem.as_bytes()),
+            IonType::List => type_qualifier_list(elem.as_sequence()),
+            IonType::SExpression => type_qualifier_sexp(elem.as_sequence()),
+            IonType::Struct => type_qualifier_struct(elem.as_struct()),
+            other => unimplemented!("TypeQualifier is not implemented for {}", other),
```

Comment:
Would it make sense to handle the dominant cases more explicitly?  E.g. if [`elem.is_null()`][1] the `Q` nibble is `0xF` full stop.  For the boolean case, deal with `true`/`false`, for the `$0` case you have the special `Q` of `0x1`, and for everything else it is `Q` of `0xF`.

Put another way, I think you can do a `const fn` to table derive the type to `0xT0` where `T` is the type code and more simply generate the low nibble `0x0Q` where `Q` is the qualifier and just bitwise-or them together.

I *think* this simplifies the logic, but maybe I am missing something.

[1]: https://docs.rs/ion-rs/0.6.0/ion_rs/value/trait.Element.html#tymethod.is_null

Reply:
That's kind of how I started and the requirement to call into specific variants (like tq-symbol for the struct case) made me rethink it. I quite like that this method "is just dispatch and no implementation".

Reply:
I guess--but it just leads to a bunch of seemingly similar code paths that are not particularly differentiated nor illuminating, but I don't feel particularly strong about it if you find this code to be more readable that the (seemingly) more concise alternative.

Reply:
I'm going to leave it as is. The ability to call into specific representations is required for structs, and I don't think we should special case symbols.